File: implementations/java-springboot/harness/src/rules/NoLoggingInDomain.java
package harness.rules;

import harness.Finding;
import harness.RuleResult;

import java.io.File;
import java.util.regex.Pattern;

import static harness.JavaFiles.collectJavaFiles;
import static harness.JavaFiles.pathContains;
import static harness.JavaFiles.readText;
import static harness.JavaFiles.relTo;

/**
 * [19] Logging is forbidden in the domain/ layer — neither {@code org.slf4j.*}, Lombok
 * {@code @Slf4j}, nor a direct {@code LoggerFactory.getLogger(...)} call may be used
 * inside domain/ (observability.md — "logging is forbidden in the Domain layer: Account
 * follows this principle — keep it that way when adding new domain methods too").
 */
public final class NoLoggingInDomain {
    private NoLoggingInDomain() {
    }

    private static final Pattern LOGGING_USAGE = Pattern.compile("org\\.slf4j|@Slf4j\\b|LoggerFactory");

    public static RuleResult check(String rootPath) {
        File root = new File(rootPath);
        RuleResult result = new RuleResult("no-logging-in-domain");
        boolean found = false;

        for (File f : collectJavaFiles(root)) {
            if (!pathContains(f, "/domain/")) continue;
            found = true;
            String rel = relTo(f, root);
            String content = readText(f);

            if (LOGGING_USAGE.matcher(content).find()) {
                result.add(Finding.fail(rel,
                    "Logging is forbidden in a domain/ class — neither slf4j/@Slf4j nor LoggerFactory may be used (observability.md)"));
            } else {
                result.add(Finding.pass(rel + " (confirmed no logging usage)"));
            }
        }

        if (!found) result.add(Finding.skip("No Java files under domain/"));
        return result;
    }
}

File: implementations/kotlin-springboot/harness/src/rules/NoLoggingInDomain.kt
package harness.rules

import harness.*
import java.io.File

// Treats all three logging APIs — direct SLF4J use, mu.KotlinLogging(microutils), and
// io.github.oshai.kotlinlogging — as signals. This repository currently uses SLF4J(observability.md),
// but adopting kotlin-logging is also under consideration, so both library names are kept on the
// blocklist.
private val LOGGING_SIGNAL = Regex(
    """import\s+org\.slf4j|import\s+mu\.KotlinLogging|import\s+io\.github\.oshai\.kotlinlogging|LoggerFactory\.getLogger|KotlinLogging\.logger""",
)

/**
 * [R8] no-logging-in-domain — domain/ does no logging (observability.md "no logging in the Domain
 * layer"). An extension of the principle that Aggregates/Domain Services must have no
 * framework/infrastructure dependency — failures are expressed as exceptions, and logging is the
 * responsibility of the Application/Infrastructure layer that catches that exception.
 */
fun checkNoLoggingInDomain(rootPath: String): RuleResult {
    val root = File(rootPath)
    val result = RuleResult("no-logging-in-domain")
    var found = false

    for (f in collectKtFiles(root)) {
        if (!f.pathContains("/domain/")) continue
        found = true
        val rel = f.relTo(root)
        val content = f.readText()
        if (LOGGING_SIGNAL.containsMatchIn(content)) {
            result.add(failFinding(rel, "logging is forbidden in the domain/ layer — SLF4J/kotlin-logging etc. may not be used (observability.md)"))
        } else {
            result.add(passFinding("$rel (no logging used)"))
        }
    }

    if (!found) result.add(skipFinding("no domain/ Kotlin files"))
    return result
}
